File: java/src/main/java/com/agentharness/orchestration/Debate.java
package com.agentharness.orchestration;

import com.agentharness.Agent;
import com.agentharness.Harness;
import com.agentharness.model.AgentInput;
import com.agentharness.model.AgentOutput;
import com.agentharness.model.DecisionAction;
import com.agentharness.model.Decisions;

import java.util.Collections;
import java.util.EnumMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Debate / Consensus orchestration (harness-protocol.md §6.4).
 *
 * <p>Multiple agents produce competing decisions over the same {@link AgentInput}; a
 * {@link ConsensusRule} reconciles them. Every participant is invoked through {@link Harness#invoke},
 * so each passes the confidence gate and tool registry individually (O-1).
 *
 * <p><b>Safety floor (invariant).</b> Consensus can never <em>raise</em> authority above the strictest
 * participant: the consensus action's severity never exceeds the strictest action any participant
 * actually proposed. Every rule chooses among proposed actions, so this holds by construction; it is
 * enforced defensively here and asserted by tests.
 */
public final class Debate {

    private final Harness harness;
    private final List<Agent> participants;
    private final ConsensusRule rule;

    public Debate(Harness harness, List<Agent> participants, ConsensusRule rule) {
        if (participants == null || participants.isEmpty()) {
            throw new IllegalArgumentException("a debate requires at least one participant");
        }
        this.harness = harness;
        this.participants = List.copyOf(participants);
        this.rule = rule == null ? ConsensusRule.SAFEST : rule;
    }

    /** A debate under the default {@code SAFEST} rule. */
    public Debate(Harness harness, List<Agent> participants) {
        this(harness, participants, ConsensusRule.SAFEST);
    }

    public DebateResult run(AgentInput request) {
        Map<String, AgentOutput> outputs = new LinkedHashMap<>();
        for (Agent p : participants) {
            outputs.put(p.name(), harness.invoke(p, request));
        }
        List<DecisionAction> actions = outputs.values().stream().map(o -> o.decision().action()).toList();

        Consensus consensus = decide(actions);
        DecisionAction action = consensus.action();

        // Safety floor (§6.4): never exceed the strictest action any participant proposed.
        DecisionAction ceiling = Decisions.reconcile(actions);
        if (Decisions.actionPrecedence(action) > Decisions.actionPrecedence(ceiling)) {
            action = ceiling;
        }

        Map<String, AgentOutput> ordered = Collections.unmodifiableMap(new LinkedHashMap<>(outputs));
        return new DebateResult(action, rule, ordered, consensus.tie());
    }

    private record Consensus(DecisionAction action, boolean tie) {
    }

    private Consensus decide(List<DecisionAction> actions) {
        if (rule == ConsensusRule.SAFEST) {
            return new Consensus(Decisions.reconcile(actions), false);
        }
        // MAJORITY: highest vote count wins; a tie for the top count resolves to DEFER.
        Map<DecisionAction, Integer> counts = new EnumMap<>(DecisionAction.class);
        for (DecisionAction a : actions) {
            counts.merge(a, 1, Integer::sum);
        }
        int top = Collections.max(counts.values());
        List<DecisionAction> winners = counts.entrySet().stream()
                .filter(e -> e.getValue() == top).map(Map.Entry::getKey).toList();
        if (winners.size() == 1) {
            return new Consensus(winners.get(0), false);
        }
        return new Consensus(DecisionAction.DEFER, true);
    }
}

File: bindings/python/src/halo_agent_harness/orchestration/debate.py
"""Debate / Consensus orchestration (spec §6.4).

Multiple agents produce competing decisions over the same ``AgentInput``; a **consensus rule**
reconciles them. Every participant is invoked through ``Harness.invoke``, so each passes the confidence
gate and tool registry individually (O-1).

Two consensus rules are provided:

- ``SAFEST`` (default) — the strictest action wins, per the Decision Hierarchy (§3.3). Conservative:
  a single ``BLOCK`` beats any number of ``ALLOW``s. This is the safe default for governance.
- ``MAJORITY`` — the most-proposed action wins; a tie resolves to ``DEFER`` (tie → human review).
  ``MAJORITY`` may *de-escalate* below the strictest proposal (that is the point of a vote); use
  ``SAFEST`` when the strictest participant must always win.

**Safety floor (invariant).** Consensus can never *raise* authority above the strictest participant:
the consensus action's severity never exceeds the strictest action any participant actually proposed.
Since every rule chooses among proposed actions, this holds by construction; it is enforced defensively
and asserted by tests.
"""

from __future__ import annotations

import enum
from collections import Counter
from dataclasses import dataclass

from ..core.agent import Agent
from ..core.harness import Harness
from ..core.model import (
    AgentInput,
    AgentOutput,
    DecisionAction,
    action_precedence,
    reconcile,
)


class ConsensusRule(enum.Enum):
    SAFEST = "safest"      # strictest action wins (Decision Hierarchy §3.3)
    MAJORITY = "majority"  # plurality; tie -> DEFER (routes to human review)


@dataclass(frozen=True)
class DebateResult:
    """The reconciled outcome of a debate."""

    consensus_action: DecisionAction
    rule: ConsensusRule
    participant_outputs: dict[str, AgentOutput]
    tie: bool = False

    @property
    def any_auto_enforced(self) -> bool:
        return any(o.decision.auto_enforced for o in self.participant_outputs.values())


class Debate:
    """Runs competing agents and reconciles their decisions with a consensus rule."""

    def __init__(
        self,
        harness: Harness,
        participants: list[Agent],
        *,
        rule: ConsensusRule = ConsensusRule.SAFEST,
    ) -> None:
        if not participants:
            raise ValueError("a debate requires at least one participant")
        self._harness = harness
        self._participants = list(participants)
        self._rule = rule

    def run(self, request: AgentInput) -> DebateResult:
        outputs = {p.name: self._harness.invoke(p, request) for p in self._participants}
        actions = [o.decision.action for o in outputs.values()]

        action, tie = self._consensus(actions)

        # Safety floor (§6.4): never exceed the strictest action any participant proposed.
        ceiling = reconcile(actions)
        if action_precedence(action) > action_precedence(ceiling):
            action = ceiling

        return DebateResult(
            consensus_action=action,
            rule=self._rule,
            participant_outputs=outputs,
            tie=tie,
        )

    def _consensus(self, actions: list[DecisionAction]) -> tuple[DecisionAction, bool]:
        if self._rule is ConsensusRule.SAFEST:
            return reconcile(actions), False

        # MAJORITY: highest vote count wins; a tie for the top count resolves to DEFER.
        counts = Counter(actions)
        top = max(counts.values())
        winners = [a for a, c in counts.items() if c == top]
        if len(winners) == 1:
            return winners[0], False
        return DecisionAction.DEFER, True
